Add reports

jorblancoa opened the pull request:
- Add support for serial/parallel
- Addressing review comments
- Add template write header functions

Commit:
Add reports
 - Integrate reportinglib

Commit:
Add support for serial/parallel
 - If MPI libraries are found use parallel, otherwise serial

Commit:
Addressing review comments
 - include cleanup
 - enum to enum class
 - catch2 as submodule
 - avoid usage of c cast
 - select serial or parallel implementation depending on MPI and HDF5 found

Fixed bug when writing to HDF5 file on serial implementation

Commit:
Added template write header functions
 - hdf5 explicit type specialization

Commit:
Added serial/parallel common interface

Commit:
Renaming
 - Adhere to SONATA
 - Coding conventions
Const correctness
Consistency on data types

Commit:
Restructure code
 - Added data, io and library folders
 - Remove factories and enums
 - Remove ReportFormat virtual class. Replace by SonataData
 - Added first exceptions

Commit:
Added logging
Added element ids from mapping

jorblancoa commented:
**TODO**
_Library_
- [X] Renaming according to SONATA nomenclature
- [X] Renaming according to HPC coding coventions
- [X] Const Correctness
- [X] Consistency on data types
- [X] Restructure code on different folders
- [X] Added logging
- [X] Support serial/parallel implementation
- [X] Remove unecesary code (non SONATA related)
- [X] Template write functions
- [X] Read element_ids from mapping instead of generating them
~- [ ] Use HighFive instead of pure hdf5 when possible : compound data type support which is not supported atm.~
- [x] Make sure all coreneuron API calls are supported in order to test the library

_Client_
- [ ] CLI arguments to test different configurations

Commit:
Fixed bug when spikes vector is empty
 - std:min/max_element() dereference end iterator
Added unsuported functions from C API

Commit:
fixup! Fixed bug when spikes vector is empty  - std:min/max_element() dereference end iterator Added unsuported functions from C API

Commit:
Record data every reporting step
 - Calculate period of recording
Adjust to coreneuron API

Commit:
Create one communicator per report
 - Avoid interlock when the same rank has different reports
Adapt to the different ways that coreneuron calls record_data
 - One chunk of gids per timestep or several timestep per chunk of gids
 - Generic way of recording and flushing the steps

Commit:
Adapt the code to the way NEURON calls reportinglib
 - NEURON: records_rec() and records_flush()
 - CoreNeuron: records_nrec() and records_end_iteration()
Add new API call to set mindelay
 - records_set_mindelay()

Commit:
Implement refresh_pointers
Avoid recording data when timestep > tend
Flushing closes the file if all steps has been recorded

Commit:
Remove update_timestamp call from record_data

Commit:
Split record_data calls
 - One receives node_ids to be recorded and the other records all the nodes

Commit:
Simplify record_data logic

Commit:
Add new API call to set the minimum steps to record
  - records_set_min_steps_to_record()
  - replace records_set_mindelay()

Commit:
Adapt unit tests to latest changes

Commit:
Added write_spikes API call

Commit:
Remove spike reports
 - Spikes are received from coreneuron ready to write

Commit:
Add tests for the write_spikes

Commit:
Install reporting lib and include files

Commit:
Change library name

Commit:
Avoid compilation failure when MPI is not found
Change log level

Commit:
Added Doxygen documentation

Commit:
Modified sonata-config.cmake to find include and lib

alkino committed:
Code refactoring
 - Clear and simplify code

tristan0x commented on src/reports/client/main.cpp:
```diff
@@ -0,0 +1,192 @@
+    std::vector<double> voltages;
+};
+
+const double dt = 0.1;
```
Shouldn't the `dt`, `tstart`, and `tstop` constants be `static` ?

tristan0x commented on src/reports/client/main.cpp:
```diff
@@ -0,0 +1,192 @@
+const double tstart = 0.0;
+const double tstop = 0.3;
+
+void generate_spikes(std::vector<uint64_t> nodeids, std::vector<double>& spike_timestamps, std::vector<int>& spike_node_ids) {
```
```suggestion
void generate_spikes(const std::vector<uint64_t>& nodeids, std::vector<double>& spike_timestamps, std::vector<int>& spike_node_ids) {
```

tristan0x commented on src/reports/client/main.cpp:
```diff
@@ -0,0 +1,192 @@
+    // Set random seed for reproducibility
+    std::srand(static_cast<unsigned int>(23487 * (seed + 1)));
+
+    std::vector<uint64_t> nodeids;
```
There is more chance to perform RVO if `nodeids` is the first variable defined in the function.

jorblancoa replied:
Didn't know that, thanks!

Commit:
Change report buffer from raw pointer to vector
Swap private and public members

tristan0x commented on src/reports/data/sonata_data.hpp:
```diff
@@ -6,9 +6,35 @@
+    void write_data();
+    void close();
+
+    bool is_due_to_report(double step);
```
Shouldn't this member method be `const`?

jorblancoa replied:
indeed

tristan0x commented on src/reports/data/sonata_data.hpp:
```diff
@@ -6,9 +6,35 @@
+    size_t get_buffer_size() const { return m_buffer_size; }
+    const std::vector<double>& get_report_buffer() const { return m_report_buffer; }
+
+    const std::vector<uint64_t>& get_node_ids() const { return m_node_ids; }
```
The `get_...` dumbs member methods should be `noexcept`

tristan0x commented on src/reports/io/hdf5_writer.cpp:
```diff
@@ -62,7 +62,7 @@ void HDF5Writer::write(double* buffer, int steps_to_write, int total_steps, int
     }*/
 
-    H5Dwrite(m_dataset, H5T_NATIVE_DOUBLE, memspace, filespace, H5P_DEFAULT, buffer);
+    H5Dwrite(m_dataset, H5T_NATIVE_DOUBLE, memspace, filespace, H5P_DEFAULT, &buffer[0]);
```
I think you are looking for `buffer.data()`

jorblancoa replied:
Yep :+1:

tristan0x commented on src/reports/io/hdf5_writer.hpp:
```diff
@@ -6,31 +6,28 @@
+    hid_t m_collective_list;
+    hid_t m_independent_list;
+
+    hsize_t m_offset[2];
```
consider using `std::array<hsize_t, 2>` instead.

tristan0x commented on src/reports/library/report.hpp:
```diff
@@ -52,6 +34,19 @@ class Report {
     void set_max_buffer_size(size_t buffer_size);
+
+  protected:
+    using nodes_t = std::map<uint64_t, Node>;
```
If the iterator ordering doesn't matter, consider using `std::unordered_map` instead.

jorblancoa replied:
I changed the report map and the communicator map to unordered.
However  in the case of the node map, although is not strictly necessary, I prefer to keep it like this for now until we release the library, because for testing/debugging it is easier to read the reports if the gids are ordered.

Commit:
Code refactoring
 - Add noexcept to getter methods
 - const correctness
 - Usage of unordered maps
 - Use vector and array .data()

alkino committed:
Node and SonataData refactoring

* Add a SomaNode to put constraint on element inside node and not inside report

* fill_data take an iterator instead of pointer

* default initializing in SonataData

* Invert if statements on sonata_data to make them more readable

alkino committed:
Add more tests (#5)

- Add tests to node and report classes

Commit:
Adapt reports to sonata target structure
 - Add sonata_report target to sonata-targets.cmake

jorblancoa committed:
API Renaming & refactoring (#4)

- Prefix sonata_ on main header file
 - Split create_report and add_node
 - Reduce parameters on some functions
 - Added namespaces to avoid name colissions

Added integration tests
 - Compare output files (compartment, soma and spikes) writen in parallel by the client app
 - Reorganise unit and integration test folders

Commit:
Apply clang-format

tristan0x commented on src/edge_index.cpp:
```diff
@@ -135,32 +130,33 @@ std::unordered_map<NodeID, RawIndex> _groupNodeRanges(const std::vector<NodeID>&
+        std::accumulate(nodeToRanges.begin(),
+                        nodeToRanges.end(),
+                        uint64_t(0),
+                        [](uint64_t total, decltype(nodeToRanges)::const_reference item) {
```
I guess you are using C++14
```suggestion
                        [](uint64_t total, const auto& item) {
```

jorblancoa replied:
Yes, however this is not from the reports part (it comes from upstream).
The changes are due to clang-format.

tristan0x commented on src/edge_index.cpp:
```diff
@@ -177,27 +173,21 @@ void _writeIndexGroup(const std::vector<NodeID>& nodeIDs, uint64_t nodeCount, Hi
-                ranges.begin(), ranges.end(),
-                std::back_inserter(secondaryIndex)
-            );
+            std::move(ranges.begin(), ranges.end(), std::back_inserter(secondaryIndex));
```
I don't understand what this instruction is doing. Honestly, I don't understand why it compiles. Did you mean `std::copy` instead?

jorblancoa replied:
As I mentioned in the previous comment, this is not from the reports.
My code is under reports folder.

Maybe I should create the PR upstream directly to show only my changes...

tristan0x commented on src/population.cpp:
```diff
@@ -21,46 +21,40 @@ namespace sonata {
 
-const Selection::Ranges& Selection::ranges() const
-{
+const Selection::Ranges& Selection::ranges() const {
```
```suggestion
const Selection::Ranges& Selection::ranges() const noexcept {
```

mgeplf committed:
added highres logo (#58)

ohm314 commented on src/reports/client/main.cpp:
```diff
@@ -0,0 +1,199 @@
+
+const static double dt = 0.1;
+const static double tstart = 0.0;
+const static double tstop = 0.3;
```
Do these need to be global vars?

jorblancoa replied:
They are used in several functions. I guess I could pass them around, but since is a test client I thought it was not so bad to keep them as global.

Ideally in the future the client could be parametrized to test different input values.

ohm314 commented on src/reports/data/sonata_data.cpp:
```diff
@@ -0,0 +1,263 @@
+
+    m_reporting_period = static_cast<int>(dt / SonataReport::m_atomic_step);
+    m_last_step_recorded = tstart / SonataReport::m_atomic_step;
+    m_last_step = tend / SonataReport::m_atomic_step;
```
why do you do a static cast on line 25 but not the others?

jorblancoa replied:
m_reporting_period is an integer, and the other 2 variables (m_last_step_recorded and m_last_step) are double

ohm314 commented on src/reports/data/sonata_data.cpp:
```diff
@@ -0,0 +1,263 @@
+                       const std::vector<int>& spike_node_ids) {
+    m_spike_timestamps = spike_timestamps;
+    m_spike_node_ids = spike_node_ids;
+    m_io_writer = std::make_unique<HDF5Writer>(report_name);
```
Why not constructed through initializer list?

ohm314 commented on src/reports/data/sonata_data.cpp:
```diff
@@ -0,0 +1,263 @@
+
+    // Calculate the timesteps that fit given the buffer size
+    {
+        int max_steps_to_write = max_buffer_size / sizeof(double) / m_total_elements;
```
```suggestion
        int max_steps_to_write = max_buffer_size / (sizeof(double) * m_total_elements);
```

ohm314 commented on src/reports/data/sonata_data.cpp:
```diff
@@ -0,0 +1,263 @@
+    }
+}
+
+bool SonataData::is_due_to_report(double step) const {
```
stupid question probably, why is step `double` ?

jorblancoa replied:
You are right, I think it makes more sense to be an int (or even better, uint64_t). But this comes from coreneuron as a double so I kept it this way.

alkino replied:
MAX_DOUBLE is far more that MAX_INT, be carefull

matz-e replied:
Can values near MAX_DOUBLE still be incremented with small values, though? I agree it would make sense to move to `u64` at some point.

ohm314 commented on src/reports/data/sonata_data.cpp:
```diff
@@ -0,0 +1,263 @@
+            m_report_buffer.size(),
+            local_position);
+    }
+    for (auto& kv : *m_nodes) {
```
Can you not get rid of some code duplication between this method and the one above?

ohm314 commented on src/reports/library/report.cpp:
```diff
@@ -0,0 +1,98 @@
+    , m_tstart(tstart)
+    , m_tend(tend)
+    , m_dt(dt) {
+    m_nodes = std::make_shared<nodes_t>();
```
You can construct this in the initializer list too

ohm314 commented on src/reports/library/report.cpp:
```diff
@@ -0,0 +1,98 @@
+    m_num_steps = static_cast<int>((tend - tstart) / dt);
+    if (std::fabs(m_num_steps * dt + tstart - tend) > std::numeric_limits<float>::epsilon()) {
+        m_num_steps++;
+    }
```
Not sure what is happening here?

jorblancoa replied:
I took it from old reportinglib.
But basically is looking if the remaining of the previous double division ((tend - tstart) / dt)) is bigger than 0.
I guess a most elegant way of putting it is:
`if (std::fabs(std::fmod((tend-tstart), dt)) > std::numeric_limits<float>::epsilon()) {`
then we do an extra step

jorblancoa replied:
Actually after testing, fmod is not good enough, sometimes doesnt behave as expected.
See https://stackoverflow.com/questions/38447548/c-fmod-returns-0-2-for-fmod0-6-0-2
The proper function to use is std::remainder(x,y)

jorblancoa replied:
this std::remainder method was still failing in some cases so I end up doing the following
`double sim_steps = (tstop - tstart)/dt;
int num_steps = static_cast<int>(std::ceil(sim_steps));
`

ohm314 commented on src/reports/library/reports.cpp:
```diff
@@ -0,0 +1,176 @@
+
+int sonata_flush(double time) {
+    if (sonata_report.is_empty()) {
+        return 0;
```
this function returns in every case 0?

ohm314 commented on src/reports/library/reports.cpp:
```diff
@@ -0,0 +1,176 @@
+
+size_t sonata_set_report_max_buffer_size_hint(char* report_name, size_t buffer_size) {
+    if (!sonata_report.report_exists(report_name)) {
+        return -1;
```
size_t is unsigned, you're returning -1... thanks @matz-e

ohm314 commented on src/reports/library/reports.cpp:
```diff
@@ -0,0 +1,176 @@
+    return 0;
+}
+
+int sonata_create_report(
```
what is exactly the semantic of the return value? why do you return 0 in any case? shouldn't you return `!= 0` if you are catching an exception?

ohm314 commented on src/reports/library/reports.cpp:
```diff
@@ -0,0 +1,176 @@
+                       uint32_t element_id,
+                       double* voltage) {
+    if (!sonata_report.report_exists(report_name)) {
+        return 0;
```
Should be also non 0 I suppose, please check the other function return values in this module.

ohm314 commented on src/reports/library/sonatareport.hpp:
```diff
@@ -0,0 +1,73 @@
+    template <typename T>
+    void apply_all(void (Report::*fun)(T), T data) {
+        std::for_each(m_reports.begin(), m_reports.end(), [&](reports_t::value_type arg) {
+            ((*(arg.second)).*fun)(data);
```
```suggestion
            std::invoke(arg.second->*fun, data)
```
I think that's a little nicer.. I've been trying to rewrite this function with `std::mem_fn` but couldn't find a good way to do this.. wondering if @tristan0x might have an idea? Just curious if this can be done more elegantly (and better type safety?) than with member function pointers?

jorblancoa replied:
Using `std::mem_fn` as discussed.

ohm314 reviewed (commented):
Huge PR :) it's in really good shape I think, but there are still a few minor things that might need cleaning up/improving.

Commit:
Added .h5 suffix to report names

mgeplf committed:
add convenience method to create a Selection of all members of a Population (#60)

Commit:
Address review comments
 - Homogenize return values on API
 - Use std::mem_fn to wrap member function pointers
 - Add tests for the number of steps in the simulation

Commit:
Better use of modern C++ / Improve Doxygen comments in C API

* Use `noexcept` specifier.
* Pass std::shared_ptr parameter as const reference.
* Add missing `override` keyword to member functions.
* Add Report::get_node const accessor.
* Document return value of C API functions.
* Remove useless IO flush.

mgeplf committed:
Make bbp::sonata::Selection comparable (#62)

* Closes #61

ohm314 reviewed (approved):
Thanks! All comments addressed.

tristan0x reviewed (approved):
LGTM

Commit:
fixup! Better use of modern C++ / Improve Doxygen comments in C API

Commit:
Revert non report changes

mgeplf committed:
Add comparison for node attributes to values (#64)

* Add comparison for node attributes to values
* This is in preparation for node_sets resolution
* fixed nodes1.h5 generation & file: enum library didn't have the same
    number of rows as the rest of the dataset
* use Catch2 as a submodule, since I needed the update version for the
  fancy TEMPLATE_TEST_CASE()

Note: It was decided that float/double comparisons were not a good idea.  See review for discussion.

Commit:
Test serial implementation
Add coverage tests

Commit:
Fix warnings

Commit:
Fix more warnings on serial implementation
Move include_directories to target_include_directories

mgeplf committed:
add coverage ci tests (#66)

* fixed missing coverage
* switch to lcov, since otherwise '50% is the new 100% coverage':
    gcovr uses a metric that includes exceptions, so unless you
    get bad_alloc to throw from every, say, std::string, you won't have
    full coverage... (see Niels Lohmann (of json c++ library fame)
    dealing with this:
    https://stackoverflow.com/questions/46575061/how-to-improve-branch-coverage-in-c

jorblancoa committed:
adapt travis file

* Add some changes to travis.yaml
  - Update gcc version to 7
  - Remove git depth
  - Add addons for to install dependencies
* Remove srand() for reproducibility across platforms
   - Adapt integration test reference output files
* Sort spikes for serial implementation

Commit:
fixup! adapt travis file

Commit:
Add coverage tests

Commit:
Change default buffer size to 4MB
Avoid writing out of bounds when steps to write are bigger than remaining steps

Commit:
fixup! Change default buffer size to 4MB Avoid writing out of bounds when steps to write are bigger than remaining steps

Commit:
Improvements by addressing comments
 - Support multipopulation.
 - Add flag to use MPI. Serial by default.
 - Remove io_writer interface.
 - Member variables consistent naming.
 - Move client app to integration tests.

jorblancoa committed:
Setup travis to test reports with/without MPI

* Setup travis to test reports with and without MPI
  - Add serial/parallel only for linux builds (OSX builds don't have HDF5-parallel)
  - Use mpich library for parallel testing
  - Fix python ci due to virtualenv update to 20.x (see https://github.com/travis-ci/cpython-builder/pull/35/files)

Commit:
Add sorting attribute for the spikes
 * by_time, by_id, none

Commit:
Addressing some comments
 * Keep consistency with int members. Changed to uint32
 * Template write functions
 * spdlog only included with reports
 * Remove using bbp::sonata namespace
 * Remove some default destructors
 * Use std::array instead of C static arrays

Commit:
More improvements
 * write_spikes is static method
 * Bring back virtual constructors to avoid clang complains
 * Avoid errors when closing the hdf5_writer more than once
 * Print log info messages only from rank 0

Commit:
Clang format

jorblancoa committed:
Collective writing

 * Every rank writes the same amount of steps (min across ranks) and write collectively
 * Update spike reports references

matz-e commented on src/reports/io/hdf5_writer.cpp:
```diff
@@ -134,12 +135,19 @@ void HDF5Writer::write(const std::string& dataset_name, const std::vector<T>& bu
 
 void HDF5Writer::close() {
-    // We close the dataset "/data" and the hdf5 file
+    // We close the dataset "/data", the spikes enum type and the hdf5 file
```
```suggestion
    // We close the dataset "/data", the spike enum type and the hdf5 file
```
?

jorblancoa committed:
Adapt to SONATA specification

 * Change report data datatype from double to float
 * Add time dataset (tstart, tstop, dt)
 * Add missing attributes (units)

Commit:
Add codeclimate config
